service/history/queueProcessor.go

```diff
@@ -0,0 +1,361 @@
+		p.NotifyNewTask() // re-enqueue the event
+		return
+	}
+
+	tasks, err := p.ackMgr.readTransferTasks()
+
+	if err != nil {
+		p.logger.Warnf("Processor unable to retrieve tasks: %v", err)
+		p.NotifyNewTask() // re-enqueue the event
+		return
+	}
+
+	if len(tasks) == 0 {
+		return
+	}
+
+	for _, tsk := range tasks {
+		tasksCh <- tsk
+	}
+
+	if len(tasks) == p.options.BatchSize {
+		// There might be more task
+		// We return now to yield, but enqueue an event to poll later
+		p.NotifyNewTask()
+	}
+
+	return
+}
+
+func (p *queueProcessorBase) taskWorker(tasksCh <-chan queueTaskInfo, workerWG *sync.WaitGroup) {
+	defer workerWG.Done()
+
+	for {
+		select {
+		case task, ok := <-tasksCh:
+			if !ok {
+				return
+			}
+
+			p.processWithRetry(task)
+		}
+	}
+}
+
+func (p *queueProcessorBase) processWithRetry(task queueTaskInfo) {
+	p.logger.Debugf("Processing task: %v, type: %v", task.GetTaskID(), task.GetTaskType())
+ProcessRetryLoop:
+	for retryCount := 1; retryCount <= 100; retryCount++ {
+		select {
+		case <-p.shutdownCh:
+			return
+		default:
+			err := p.processor.Process(task)
+
+			if err != nil {
+				logging.LogTaskProcessingFailedEvent(p.logger, task.GetTaskID(), task.GetTaskType(), err)
+				backoff := time.Duration(retryCount * 100)
+				time.Sleep(backoff * time.Millisecond)
+				continue ProcessRetryLoop
+			}
+
+			p.ackMgr.completeTask(task.GetTaskID())
+			return
+		}
+	}
+
+	// All attempts to process transfer task failed.  We won't be able to move the ackLevel so panic
+	logging.LogOperationPanicEvent(p.logger,
+		fmt.Sprintf("Retry count exceeded for taskID: %v", task.GetTaskID()), nil)
+}
+
+func (a *ackManager) readTransferTasks() ([]queueTaskInfo, error) {
+	a.RLock()
+	rLevel := a.readLevel
+	a.RUnlock()
+
+	var tasks []queueTaskInfo
+	op := func() error {
+		var err error
+		tasks, err = a.processor.ReadTasks(rLevel)
+		return err
+	}
+
+	err := backoff.Retry(op, persistenceOperationRetryPolicy, common.IsPersistenceTransientError)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(tasks) == 0 {
+		return tasks, nil
+	}
+
+	a.Lock()
+	for _, task := range tasks {
+		if a.readLevel >= task.GetTaskID() {
+			a.logger.Fatalf("Next task ID is less than current read level.  TaskID: %v, ReadLevel: %v", task.GetTaskID(),
+				a.readLevel)
+		}
+		a.logger.Debugf("Moving read level: %v", task.GetTaskID())
+		a.readLevel = task.GetTaskID()
+		a.outstandingTasks[a.readLevel] = false
+	}
+	a.Unlock()
+
+	return tasks, nil
+}
+
+func (a *ackManager) completeTask(taskID int64) {
+	a.Lock()
+	if _, ok := a.outstandingTasks[taskID]; ok {
+		a.outstandingTasks[taskID] = true
+	}
+	a.Unlock()
+}
+
+func (a *ackManager) updateAckLevel() {
+	a.metricsClient.IncCounter(a.options.MetricScope, metrics.AckLevelUpdateCounter)
+	initialAckLevel := a.ackLevel
+	updatedAckLevel := a.ackLevel
+	a.Lock()
+MoveAckLevelLoop:
+	for current := a.ackLevel + 1; current <= a.readLevel; current++ {
+		if acked, ok := a.outstandingTasks[current]; ok {
```

Comment:
we should probably panic if !ok, i.e. key not found.

Reply:
Well for now I ported over the existing logic.  Let me add a comment address what happens if !ok.